internal/ethereum/confirmation_reconciler.go

```diff
@@ -52,71 +59,238 @@ func (bl *blockListener) reconcileConfirmationsForTransaction(ctx context.Contex
+		ParentHash:  freshBlockInfo.ParentHash.String(),
 	}
 
-	// Validate existing confirmations and fill gaps in the confirmation queue
-	var confirmations []*ffcapi.MinimalBlockInfo
-	var newFork bool
-	confirmations, newFork, err = bl.checkAndFillInGap(ctx, newConfirmationsWithoutTxBlock, existingConfirmations, txBlockInfo, targetConfirmationCount, lastValidatedBlock)
+	// Validate parent-child relationship
+	if !fetchedBlock.IsParentOf(s.lateList[0]) {
+		// most likely explanation of this is an unstable chain
+		return i18n.NewError(ctx, msgs.MsgFailedToBuildConfirmationQueue)
+	}
+
+	// Prepend fetched block to confirmation queue
+	s.lateList = append([]*ffcapi.MinimalBlockInfo{fetchedBlock}, s.lateList...)
+	return nil
+}
+
+func (s *splice) removeBrokenLink() {
+	if s.hasGap() {
+		// nothing to remove if there is a gap
+		return
+	}
+	// remove the last block from the early list because it is not the parent of the first block in the late list and we have higher confidence in the late list
+	s.earlyList = s.earlyList[:len(s.earlyList)-1]
+
+}
+
+func (s *splice) toSingleLinkedList() []*ffcapi.MinimalBlockInfo {
+	if s.hasGap() {
+		return nil
+	}
+	if len(s.earlyList) == 0 {
+		return s.lateList
+	}
+	if len(s.lateList) == 0 {
+		return s.earlyList
+	}
+	if s.earlyList[len(s.earlyList)-1].IsParentOf(s.lateList[0]) {
+		return append(s.earlyList, s.lateList...)
+	}
+	// cannot be linked because the last block in the early list is not the parent of the first block in the late list
+	return nil
+
+}
+
+func createEarlyList(ctx context.Context, existingConfirmations []*ffcapi.MinimalBlockInfo, txBlockInfo *ffcapi.MinimalBlockInfo, reconcileResult *ffcapi.ConfirmationUpdateResult) (earlyList []*ffcapi.MinimalBlockInfo) {
+	if len(existingConfirmations) > 0 && !existingConfirmations[0].Equal(txBlockInfo) {
+		// otherwise we discard the existing confirmations queue
+		reconcileResult.NewFork = true
+	} else {
+		earlyList = existingConfirmations
+	}
+
+	if len(existingConfirmations) == 0 {
+		// either because this is the first time we are reconciling this transaction or because we just discarded the existing confirmations queue
+		earlyList = []*ffcapi.MinimalBlockInfo{txBlockInfo}
+	}
+	return earlyList
+}
+
+func createLateList(ctx context.Context, txBlockInfo *ffcapi.MinimalBlockInfo, targetConfirmationCount uint64, reconcileResult *ffcapi.ConfirmationUpdateResult, blockListener *blockListener) (lateList []*ffcapi.MinimalBlockInfo, err error) {
+	lateList, err = blockListener.buildConfirmationQueueUsingInMemoryPartialChain(ctx, txBlockInfo, targetConfirmationCount)
 	if err != nil {
 		return nil, err
 	}
-	reconcileResult.NewFork = newFork
-	reconcileResult.Confirmations = confirmations
-	reconcileResult.Confirmed = uint64(len(confirmations)) >= targetConfirmationCount+1
-	return reconcileResult, err
+
+	// If the late list is empty, it may be because the chain has moved on so far and the transaction is so old that
```

Comment:
```suggestion
    // buildConfirmationQueueUsingInMemoryPartialChain will return an error if the target block is after the last block in the in-memory partial chain
	// Therefore, if the late list is empty, it must be because the chain has moved on so far, and the transaction is so old that
```